include/bout/region.hxx

```diff
@@ -22,36 +22,74 @@
+///
+/// The `Region` class keeps hold of a set of indices which are used
+/// to index a Field. Here, a Field refers to either a Field2D or a
+/// Field3D. These indices can either be used directly, or blocks of
+/// contiguous indices may be used instead, which allows OpenMP
+/// parallelisation.
+///
+/// The BLOCK_REGION_LOOP helper macro is provided as a replacement
+/// for for-loops.
+///
+/// Separate index classes are available for Field2Ds (Ind2D) and
+/// Field3Ds (Ind3D). Note that while an Ind3D can be used to index a
+/// Field2D, an Ind2D cannot be used to index a Field3D. This is
+/// because an Ind2D essentially doesn't keep track of the
+/// z-dimension.
+
 #ifndef __REGION_H__
 #define __REGION_H__
 
+#include <algorithm>
 #include <type_traits>
 #include <utility>
 #include <vector>
 
+#include "bout/assert.hxx"
 #include "bout/openmpwrap.hxx"
 
-/*
- * The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
- * performance on specific hardware. It determines what the largest
- * contiguous block size can be. As we hope the compiler will vectorise
- * the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE
- * is likely related to the vector size etc.
- */
+/// The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
+/// performance on specific hardware. It determines what the largest
+/// contiguous block size can be. As we hope the compiler will vectorise
+/// the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE
+/// is likely related to the vector size etc.
 #ifndef MAXREGIONBLOCKSIZE
 #define MAXREGIONBLOCKSIZE 64
 #endif
 #if MAXREGIONBLOCKSIZE <= 0
 #error "In region.hxx must set MAXREGIONBLOCKSIZE > 0"
 #endif
 
-// The following loops vectorise well well index is type int, needs testing
-// with current approach where index is Ind2D or ind3D
+/// Helper macros for iterating over a Region making use of the
+/// contiguous blocks of indices
+///
+/// @param[in] region An already existing Region
+/// @param[in] index  The name of the index variable to use in the loop
+/// The rest of the arguments form the loop body.
+///
+/// The following loops vectorise well when index is type int, needs testing
+/// with current approach where index is Ind2D or ind3D
+///
+/// Example
+/// -------
+/// The following for-loop:
+///
+///     for (auto index = begin(region); index < end(region); ++index) {
+///        A[index] = B[index] + C[index];
+///     }
+///
+/// can be converted to a block region loop like so:
+///
+///     BLOCK_REGION_LOOP(region, index,
+///        A[index] = B[index] + C[index];
+///     )
 #define BLOCK_REGION_LOOP_SERIAL(region, index, ...)                                     \
   {                                                                                      \
     const auto blocks = region.getBlocks();                                              \
     for (auto block = blocks.begin(); block < blocks.end(); ++block) {                   \
```

Comment:
Should this be `!=` rather than `<`? I'm not sure if there is any performance difference (maybe one instruction), but != seems to be standard. I guess it has fewer requirements on the comparison operator.

Reply:
The comments on https://stackoverflow.com/a/17853547 suggests `!=` may not work. Certainly all examples I found used `<`, should be easy enough to test though!.